sys/src/9/port/sysfile.c

```diff
@@ -847,7 +843,7 @@ write(int fd, void *p, int32_t n, int64_t off)
 
 	n = r;
 
-	if(sequential){	/* use and maintain channel's offset */
+	if(dopwrite){	/* use and maintain channel's offset */
```

Comment:
Why not avoid advancing c->offset until after r->dev->write returns?

Reply:
Answering my own question...If I had to guess:

The authors increment offset so that multiple system calls writing to the chan at the same time don't overlap. However, there's a time-of-check, time-of-use bug here (I think) with setting `off` when off==-1LL.

Probably the best thing is to add, `int pread = (off != -1LL);` at the beginning and test on that to set off from c->offset and increment c->offset before c->dev->write.